tensorboard/pip_package/build_pip_package.sh

```diff
@@ -29,93 +27,113 @@ Arguments:
 EOF
 }
 
-if [ $# -ne 1 ]; then
-  usage 2>&1
-  exit 1
-fi
-output="$1"
-
-if [ -z "${RUNFILES+set}" ]; then
-  RUNFILES="$(CDPATH="" cd -- "$0.runfiles" && pwd)"
-fi
-
-if [ "$(uname)" = "Darwin" ]; then
-  sedi="sed -i ''"
-else
-  sedi="sed -i"
-fi
-
-command -v virtualenv >/dev/null
-[ -d "${RUNFILES}" ]
-
-if [ "$(uname)" = "Darwin" ]; then
-  workdir="$(mktemp -d -t tensorboard-pip)"
-else
-  workdir="$(mktemp -d -p /tmp -t tensorboard-pip.XXXXXXXXXX)"
-fi
-original_wd="${PWD}"
-cd "${workdir}"
-
-cleanup() {
-  rm -r "${workdir}"
+main() {
+  if [ $# -ne 1 ]; then
+    usage 2>&1
+    return 1
+  fi
+  output="$1"
+
+  if [ -z "${RUNFILES+set}" ]; then
+    RUNFILES="$(CDPATH="" cd -- "$0.runfiles" && pwd)"
+  fi
+
+  if [ "$(uname)" = "Darwin" ]; then
+    workdir="$(mktemp -d -t tensorboard-pip)"
+  else
+    workdir="$(mktemp -d -p /tmp -t tensorboard-pip.XXXXXXXXXX)"
+  fi
+  original_wd="${PWD}"
+  cd "${workdir}" || return 2
+
+  cleanup() {
+    rm -r "${workdir}"
+  }
+  trap cleanup EXIT
+
+  log_file="${workdir}/log"
+  suppress_log() {
+    rm "${workdir}/log" || true
+  }
+
+  build 3>&1 4>&2 1>"${log_file}" 2>&1
```

Comment:
It seems like we could just do redirection on 1 and 2 and skip the slightly more unusual use of 3 and 4 if we were willing to just say that there isn't any way to write to the original stdout and stderr from within `build` - we don't seem to be using `3` at all, and the one use of `4` (to emit the no-such-directory error) seems like we could just write that to the redirected stderr and print a more verbose log message.  (In which case we don't need to implement `suppress_log` either, and we could change the `if ! [ -d "${output}" ]` check to just be an unadorned `[ -d "${output}" ]` like we do earlier for `RUNFILES`.)

Reply:
Sure thing: removed the redirections in favor of a noisier log. And indeed it’s really not that bad:

```
+ case "${output}" in
+ '[' -d /tmp/nope ']'
+ printf 'fatal: no such output directory: %s\n' /tmp/nope
fatal: no such output directory: /tmp/nope
+ return 1
```

I’ve kept the explicit `! [ -d "${output}" ]` because this is a usage
error and imho deserves a nicer message (and it’s not hurting
readability or anything).